src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -285,71 +286,52 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         }
     }
 
-    /// Check if a borrow location is within a loop.
-    fn is_borrow_location_in_loop(
+    /// true if the `borrow_location` can reach the `use_location` without passing through any
+    /// backedges.
+    fn can_reach_ignoring_backedges(
         &self,
         borrow_location: Location,
+        use_location: Location,
     ) -> bool {
-        let mut visited_locations = Vec::new();
-        let mut pending_locations = vec![ borrow_location ];
-        debug!("is_in_loop: borrow_location={:?}", borrow_location);
+        let mut visited_locations = FxHashSet::default();
+        let mut pending_locations = vec![borrow_location];
+        debug!(
+            "can_reach_ignoring_backedges: borrow_location={:?}",
+            borrow_location
+        );
```

Comment:
btw we should add `borrow_location` to `visited_locations` here, but not a big thing